File: lib/domain/settings/app_settings.dart
// VERIFY: OQ-B1 — confirm each key string against the upstream GNOME Buffer
// gschema before TASK-13 (SharedPreferencesSettingsRepository) is committed.
// A mismatch means stored values are never found on-device.
import 'package:freezed_annotation/freezed_annotation.dart';

part 'app_settings.freezed.dart';

/// Theme-mode preference (FR-M6-01 / EC-01).
///
/// - [follow]: honour the OS dark/light setting (default).
/// - [light]: always Light theme.
/// - [dark]: always Dark theme.
enum AppColorScheme { follow, light, dark }

@freezed
class AppSettings with _$AppSettings {
  // Private constructor required for custom getters on freezed classes.
  const AppSettings._();

  const factory AppSettings({
    @Default(true) bool useMonospaceFont,
    @Default(true) bool spellingEnabled,
    @Default(true) bool emergencyRecoveryEnabled,
    @Default(true) bool lineLengthEnabled,
    @Default(AppColorScheme.follow) AppColorScheme colorScheme,
    // FR-M7-01: default index 8 → 14pt (slotList[8] == 14).
    @Default(8) int fontSizeIndex,
  }) = _AppSettings;

  // ---------------------------------------------------------------------------
  // 21-slot font-size table — FR-M7-01
  // Index 8 (default) → 14pt.  Strictly ascending; no duplicates.
  // ---------------------------------------------------------------------------
  static const List<int> slotList = [
    6,
    7,
    8,
    9,
    10,
    11,
    12,
    13,
    14,
    15,
    16,
    17,
    18,
    20,
    22,
    24,
    26,
    28,
    30,
    34,
    38,
  ];

  // ---------------------------------------------------------------------------
  // Integer-proxy getters — FR-12
  // The boolean toggle is the single mutation point (EC-06); the integer value
  // is always derived and never stored independently.
  // ---------------------------------------------------------------------------

  /// Number of emergency-recovery files to retain.
  /// Derives from [emergencyRecoveryEnabled]: true → 10, false → 0.
  int get emergencyRecoveryFiles => emergencyRecoveryEnabled ? 10 : 0;

  /// Maximum line length in characters.
  /// Derives from [lineLengthEnabled]: true → 800 (upstream on-value), false → 100000 (effectively unlimited).
  int get lineLength => lineLengthEnabled ? 800 : 100000;

  // ---------------------------------------------------------------------------
  // Typography derived getter — FR-M7-01
  // ---------------------------------------------------------------------------

  /// Current font size in points derived from [fontSizeIndex].
  double get fontSizePt => slotList[fontSizeIndex].toDouble();

  // ---------------------------------------------------------------------------
  // Pure clampable verbs — FR-M7-03, FR-M7-04
  // Each verb short-circuits (returns `this`) when the value is unchanged,
  // because freezed's copyWith always produces a new instance even for equal
  // values, breaking `identical(this)` no-op assertions.
  // ---------------------------------------------------------------------------

  /// Sets [fontSizeIndex] to [index], clamped to [0, slotList.length-1].
  /// Returns `this` when the clamped value equals the current index (no-op).
  AppSettings setFontSizeIndex(int index) {
    final clamped = index.clamp(0, slotList.length - 1);
    if (clamped == fontSizeIndex) return this;
    return copyWith(fontSizeIndex: clamped);
  }

  /// Sets [useMonospaceFont] to [value].
  /// Returns `this` when the value is already [value] (idempotent, NOT a toggle).
  AppSettings setUseMonospaceFont(bool value) {
    if (useMonospaceFont == value) return this;
    return copyWith(useMonospaceFont: value);
  }

  // ---------------------------------------------------------------------------
  // Key strings — must match the upstream GNOME Buffer gschema verbatim.
  // See VERIFY: OQ-B1 above.
  // ---------------------------------------------------------------------------

  static const String kUseMonospaceFont = 'use-monospace-font';
  static const String kSpellingEnabled = 'check-spelling';
  static const String kEmergencyRecoveryEnabled = 'save-emergency-files';
  static const String kEmergencyRecoveryFiles = 'emergency-recovery-files';
  static const String kLineLength = 'line-length';

  /// Key for the [colorScheme] setting.
  /// Pinned as 'color-scheme' per OQ-M6-05 (NOT 'style-variant').
  static const String kColorScheme = 'color-scheme';

  /// Key for the [fontSizeIndex] setting — FR-M7-01 / OQ-M7-03.
  static const String kFontSize = 'font-size';
}

File: shared/src/commonMain/kotlin/com/paolosantucci/foglietto/shared/settings/AppSettings.kt
package com.paolosantucci.foglietto.shared.settings

/**
 * Theme-mode preference.
 *
 * Lower-case entries so `.name` round-trips the wire value stored under
 * [AppSettings.KEY_COLOR_SCHEME] (i.e. `"follow"`, `"light"`, `"dark"`).
 *
 * Spec §5.1.2 / FR-19. Parse via explicit `when` — never via `valueOf`/`enumValueOf`
 * (which throw on an unknown value). That defensive parse is SettingsRepositoryImpl's
 * concern (FR-22, TASK-06).
 */
enum class AppColorScheme { follow, light, dark }

/**
 * Immutable settings model — MVP-trimmed port of the Dart `AppSettings` class.
 *
 * Trimmed surface (OQ-02 drop-list — the following DO NOT exist by construction):
 *   - useMonospaceFont, spellingEnabled, emergencyRecoveryEnabled, lineLengthEnabled
 *   - emergencyRecoveryFiles, lineLength (integer-proxy getters)
 *   - setUseMonospaceFont
 *   - keys: use-monospace-font, check-spelling, save-emergency-files,
 *            emergency-recovery-files, line-length
 *
 * @property colorScheme Theme-mode preference; default [AppColorScheme.follow].
 * @property fontSizeIndex Index into [slotList]; default 8 → 14pt.
 *
 * Spec refs: FR-19, FR-20, FR-21, §5.1.2; S-C1, S-I3 (assessment).
 */
data class AppSettings(
    val colorScheme: AppColorScheme = AppColorScheme.follow,
    val fontSizeIndex: Int = 8,
) {

    /**
     * Current font size in points derived from [fontSizeIndex].
     *
     * Computed val (never stored). Spec §5.1.2 declares this as `Int`; the Dart
     * oracle returns `double` for Dart's type system — the Kotlin port follows
     * the spec and returns `Int`.
     */
    val fontSizePt: Int get() = slotList[fontSizeIndex]

    /**
     * Returns a copy of this settings with [fontSizeIndex] set to [index],
     * clamped to `[0, slotList.lastIndex]`.
     *
     * Identity-stable (S-C1): when the clamped value equals the current
     * [fontSizeIndex], returns `this` — NOT a new `copy()` — so that callers
     * can test `assertSame` to catch regressions where an unconditional
     * `copy()` would produce a distinct object even on a no-op.
     *
     * The identity-stable guard (`if (clamped == fontSizeIndex) return this`)
     * MUST appear BEFORE `copy(...)`.
     *
     * Spec FR-20, S-C1, S-I3.
     */
    fun setFontSizeIndex(index: Int): AppSettings {
        val clamped = index.coerceIn(0, slotList.lastIndex)
        if (clamped == fontSizeIndex) return this
        return copy(fontSizeIndex = clamped)
    }

    /**
     * Returns a copy of this settings with [colorScheme] set to [scheme],
     * preserving [fontSizeIndex] verbatim.
     *
     * Identity-stable (S-C1, mirroring [setFontSizeIndex]): when [scheme]
     * equals the current [colorScheme], returns `this` — NOT a new `copy()`.
     * This allows callers (e.g. the theme picker) to use value-equality as a
     * no-op guard (FR-10) without allocating a new instance.
     *
     * Non-suspend. Additive — no existing member changed or removed.
     * Platform-neutral (commonMain) — no Apple/JVM platform dependency.
     *
     * Spec FR-04, FR-24, NFR-08; closes CM-2. Contract §5.1.a.
     */
    fun setColorScheme(scheme: AppColorScheme): AppSettings {
        if (scheme == colorScheme) return this
        return copy(colorScheme = scheme)
    }

    companion object {
        /**
         * 21-slot font-size table.
         *
         * Declared `val` (NOT `const`) because `const` requires a `List<Int>` to
         * be a compile-time constant, which is not supported for collection types
         * in Kotlin. The plan explicitly requires `val`, not `const`.
         *
         * Invariants (FR-21, S-I3):
         *   - length == 21
         *   - slotList[8] == 14  (the default index)
         *   - strictly ascending
         *   - no duplicates
         *   - slotList[0] == 6, slotList[20] == 38
         */
        val slotList: List<Int> = listOf(
            6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 20, 22, 24, 26, 28, 30, 34, 38
        )

        /** Key for the [colorScheme] setting in the multiplatform-settings store. */
        const val KEY_COLOR_SCHEME: String = "color-scheme"

        /** Key for the [fontSizeIndex] setting in the multiplatform-settings store. */
        const val KEY_FONT_SIZE: String = "font-size"
    }
}
